File: packages/graph-core/src/validator.ts
import { GraphDefinitionSchema } from "./schemas";
import {
  GRAPH_VALIDATION_ERROR_CODES,
  GraphValidationError,
  type GraphValidationErrorCode
} from "./errors";
import type { GraphDefinition } from "./types";

const isNonEmptyCondition = (value: string | undefined): value is string =>
  typeof value === "string" && value.trim().length > 0;

const createError = (
  code: GraphValidationErrorCode,
  message: string,
  path: (string | number)[] = []
): GraphValidationError => new GraphValidationError(code, message, path);

export const validateGraph = (def: GraphDefinition): GraphValidationError[] => {
  const errors: GraphValidationError[] = [];
  const schemaResult = GraphDefinitionSchema.safeParse(def);

  if (!schemaResult.success) {
    for (const issue of schemaResult.error.issues) {
      if (issue.path.includes("condition")) {
        errors.push(
          createError(
            GRAPH_VALIDATION_ERROR_CODES.INVALID_CONDITION_FORMAT,
            issue.message,
            issue.path
          )
        );
      }
    }
  }

  const nodeIds = new Set<string>();
  for (let i = 0; i < def.nodes.length; i += 1) {
    const node = def.nodes[i];
    const nodeId = node?.id;
    if (node === undefined || nodeId === undefined) {
      continue;
    }

    if (nodeIds.has(nodeId)) {
      errors.push(
        createError(
          GRAPH_VALIDATION_ERROR_CODES.DUPLICATE_NODE_ID,
          `Duplicate node id '${nodeId}'.`,
          ["nodes", i, "id"]
        )
      );
      continue;
    }

    nodeIds.add(nodeId);

    if (node.type === "subgraph" && node.subgraphId === undefined) {
      errors.push(
        createError(
          GRAPH_VALIDATION_ERROR_CODES.INVALID_EDGE_REFERENCE,
          `Subgraph node '${nodeId}' must define subgraphId.`,
          ["nodes", i, "subgraphId"]
        )
      );
    }

    const inputMapping = node.inputMapping;
    if (inputMapping !== undefined) {
      const isValidInputMapping = Object.entries(inputMapping).every(
        ([targetKey, sourceKey]) => typeof targetKey === "string" && typeof sourceKey === "string"
      );
      if (!isValidInputMapping) {
        errors.push(
          createError(
            GRAPH_VALIDATION_ERROR_CODES.INVALID_CONDITION_FORMAT,
            `Node '${nodeId}' inputMapping must be string-to-string.`,
            ["nodes", i, "inputMapping"]
          )
        );
      }
    }

    const outputMapping = node.outputMapping;
    if (outputMapping !== undefined) {
      const isValidOutputMapping = Object.entries(outputMapping).every(
        ([targetKey, sourceKey]) => typeof targetKey === "string" && typeof sourceKey === "string"
      );
      if (!isValidOutputMapping) {
        errors.push(
          createError(
            GRAPH_VALIDATION_ERROR_CODES.INVALID_CONDITION_FORMAT,
            `Node '${nodeId}' outputMapping must be string-to-string.`,
            ["nodes", i, "outputMapping"]
          )
        );
      }
    }
  }

  const edgeIds = new Set<string>();
  for (let i = 0; i < def.edges.length; i += 1) {
    const edge = def.edges[i];
    if (edge === undefined) {
      continue;
    }

    if (edgeIds.has(edge.id)) {
      errors.push(
        createError(
          GRAPH_VALIDATION_ERROR_CODES.DUPLICATE_EDGE_ID,
          `Duplicate edge id '${edge.id}'.`,
          ["edges", i, "id"]
        )
      );
    } else {
      edgeIds.add(edge.id);
    }

    if (!nodeIds.has(edge.from)) {
      errors.push(
        createError(
          GRAPH_VALIDATION_ERROR_CODES.INVALID_EDGE_REFERENCE,
          `Edge '${edge.id}' references unknown source node '${edge.from}'.`,
          ["edges", i, "from"]
        )
      );
    }

    if (!nodeIds.has(edge.to)) {
      errors.push(
        createError(
          GRAPH_VALIDATION_ERROR_CODES.INVALID_EDGE_REFERENCE,
          `Edge '${edge.id}' references unknown target node '${edge.to}'.`,
          ["edges", i, "to"]
        )
      );
    }

    if (edge.type === "conditional" && !isNonEmptyCondition(edge.condition)) {
      errors.push(
        createError(
          GRAPH_VALIDATION_ERROR_CODES.INVALID_CONDITION_FORMAT,
          `Conditional edge '${edge.id}' must declare a non-empty named condition.`,
          ["edges", i, "condition"]
        )
      );
    }

    if (edge.condition !== undefined && !isNonEmptyCondition(edge.condition)) {
      errors.push(
        createError(
          GRAPH_VALIDATION_ERROR_CODES.INVALID_CONDITION_FORMAT,
          `Edge '${edge.id}' has an invalid condition format.`,
          ["edges", i, "condition"]
        )
      );
    }
  }

  if (!nodeIds.has(def.entryNodeId)) {
    errors.push(
      createError(
        GRAPH_VALIDATION_ERROR_CODES.MISSING_ENTRY_NODE,
        `Entry node '${def.entryNodeId}' does not exist in nodes.`,
        ["entryNodeId"]
      )
    );
  }

  for (let i = 0; i < def.nodes.length; i += 1) {
    const node = def.nodes[i];
    if (node?.fanOut === undefined) {
      continue;
    }
    for (const fanNodeId of node.fanOut.parallelTo) {
      if (!nodeIds.has(fanNodeId)) {
        errors.push(
          createError(
            GRAPH_VALIDATION_ERROR_CODES.INVALID_EDGE_REFERENCE,
            `Fan-out node '${node.id}' references unknown parallel node '${fanNodeId}'.`,
            ["nodes", i, "fanOut", "parallelTo"]
          )
        );
      }
    }
    if (!nodeIds.has(node.fanOut.joinAt)) {
      errors.push(
        createError(
          GRAPH_VALIDATION_ERROR_CODES.INVALID_EDGE_REFERENCE,
          `Fan-out node '${node.id}' references unknown join node '${node.fanOut.joinAt}'.`,
          ["nodes", i, "fanOut", "joinAt"]
        )
      );
    }
  }

  const hasCycle = (() => {
    const adjacency = new Map<string, string[]>();
    for (const nodeId of nodeIds) {
      adjacency.set(nodeId, []);
    }
    for (const edge of def.edges) {
      const from = adjacency.get(edge.from);
      if (from !== undefined) {
        from.push(edge.to);
      }
    }
    const visiting = new Set<string>();
    const visited = new Set<string>();
    const dfs = (nodeId: string): boolean => {
      if (visiting.has(nodeId)) {
        return true;
      }
      if (visited.has(nodeId)) {
        return false;
      }
      visiting.add(nodeId);
      for (const to of adjacency.get(nodeId) ?? []) {
        if (dfs(to)) {
          return true;
        }
      }
      visiting.delete(nodeId);
      visited.add(nodeId);
      return false;
    };
    for (const nodeId of nodeIds) {
      if (dfs(nodeId)) {
        return true;
      }
    }
    return false;
  })();

  if (hasCycle && def.recursionLimit === undefined) {
    errors.push(
      createError(
        GRAPH_VALIDATION_ERROR_CODES.CYCLE_DETECTED,
        "Cycle detected and recursionLimit is missing. Set recursionLimit to acknowledge cyclic execution.",
        ["recursionLimit"]
      )
    );
  }

  return errors;
};

File: crates/graph-core/src/validator.rs
//! Structural validation of a [`GraphDefinition`]. Mirrors the TS `validateGraph`:
//! returns every problem found rather than failing on the first.

use std::collections::HashSet;

use crate::error::{ValidationError, ValidationErrorCode};
use crate::types::{EdgeType, GraphDefinition};

/// Validate a graph definition. An empty vec means the graph is structurally sound.
pub fn validate_graph(def: &GraphDefinition) -> Vec<ValidationError> {
    let mut errors = Vec::new();

    let mut node_ids: HashSet<&str> = HashSet::new();
    for node in &def.nodes {
        if !node_ids.insert(node.id.as_str()) {
            errors.push(ValidationError::new(
                ValidationErrorCode::DuplicateNodeId,
                format!("Duplicate node id '{}'.", node.id),
                vec![node.id.0.clone()],
            ));
        }
    }

    let mut edge_ids: HashSet<&str> = HashSet::new();
    for edge in &def.edges {
        if !edge_ids.insert(edge.id.as_str()) {
            errors.push(ValidationError::new(
                ValidationErrorCode::DuplicateEdgeId,
                format!("Duplicate edge id '{}'.", edge.id),
                vec![edge.id.0.clone()],
            ));
        }
        if !node_ids.contains(edge.from.as_str()) {
            errors.push(ValidationError::new(
                ValidationErrorCode::InvalidEdgeReference,
                format!(
                    "Edge '{}' references unknown node '{}'.",
                    edge.id, edge.from
                ),
                vec![edge.id.0.clone()],
            ));
        }
        if !node_ids.contains(edge.to.as_str()) {
            errors.push(ValidationError::new(
                ValidationErrorCode::InvalidEdgeReference,
                format!("Edge '{}' references unknown node '{}'.", edge.id, edge.to),
                vec![edge.id.0.clone()],
            ));
        }
        if edge.edge_type == EdgeType::Conditional {
            let valid = edge
                .condition
                .as_ref()
                .map(|condition| !condition.trim().is_empty())
                .unwrap_or(false);
            if !valid {
                errors.push(ValidationError::new(
                    ValidationErrorCode::InvalidConditionFormat,
                    format!(
                        "Conditional edge '{}' requires a non-empty named condition.",
                        edge.id
                    ),
                    vec![edge.id.0.clone()],
                ));
            }
        }
    }

    if !node_ids.contains(def.entry_node_id.as_str()) {
        errors.push(ValidationError::new(
            ValidationErrorCode::MissingEntryNode,
            format!(
                "Entry node '{}' is not declared in the graph.",
                def.entry_node_id
            ),
            vec![def.entry_node_id.0.clone()],
        ));
    }

    errors
}

#[cfg(test)]
mod tests {
    use std::collections::BTreeMap;

    use crate::ids::{EdgeId, GraphId, NodeId};
    use crate::types::{
        ChannelDefinition, ChannelReducer, EdgeDefinition, EdgeType, GraphDefinition,
        NodeDefinition, NodeType,
    };

    use super::*;

    fn node(id: &str, node_type: NodeType) -> NodeDefinition {
        NodeDefinition {
            id: NodeId::from(id),
            node_type,
            label: id.to_owned(),
            subgraph_id: None,
            input_mapping: None,
            output_mapping: None,
            fan_out: None,
            retry_policy: None,
            metadata: None,
        }
    }

    fn edge(
        id: &str,
        from: &str,
        to: &str,
        edge_type: EdgeType,
        condition: Option<&str>,
    ) -> EdgeDefinition {
        EdgeDefinition {
            id: EdgeId::from(id),
            from: NodeId::from(from),
            to: NodeId::from(to),
            edge_type,
            condition: condition.map(|c| c.to_owned()),
        }
    }

    fn graph(
        nodes: Vec<NodeDefinition>,
        edges: Vec<EdgeDefinition>,
        entry: &str,
    ) -> GraphDefinition {
        let mut channels = BTreeMap::new();
        channels.insert(
            "count".to_owned(),
            ChannelDefinition {
                channel_type: "number".to_owned(),
                reducer: ChannelReducer::Replace,
                default: None,
            },
        );
        GraphDefinition {
            id: GraphId::from("g"),
            version: "0.0.0".to_owned(),
            name: "g".to_owned(),
            recursion_limit: None,
            channels,
            nodes,
            edges,
            entry_node_id: NodeId::from(entry),
            metadata: None,
        }
    }

    #[test]
    fn accepts_a_well_formed_graph() {
        let def = graph(
            vec![node("a", NodeType::Action), node("b", NodeType::Action)],
            vec![edge("e1", "a", "b", EdgeType::Default, None)],
            "a",
        );
        assert!(validate_graph(&def).is_empty());
    }

    #[test]
    fn flags_a_dangling_edge_reference() {
        let def = graph(
            vec![node("a", NodeType::Action)],
            vec![edge("e1", "a", "ghost", EdgeType::Default, None)],
            "a",
        );
        let errors = validate_graph(&def);
        assert!(errors
            .iter()
            .any(|e| e.code == ValidationErrorCode::InvalidEdgeReference));
    }

    #[test]
    fn flags_a_missing_entry_node() {
        let def = graph(vec![node("a", NodeType::Action)], vec![], "nope");
        let errors = validate_graph(&def);
        assert!(errors
            .iter()
            .any(|e| e.code == ValidationErrorCode::MissingEntryNode));
    }

    #[test]
    fn flags_a_conditional_edge_without_a_condition() {
        let def = graph(
            vec![node("a", NodeType::Action), node("b", NodeType::Action)],
            vec![edge("e1", "a", "b", EdgeType::Conditional, None)],
            "a",
        );
        let errors = validate_graph(&def);
        assert!(errors
            .iter()
            .any(|e| e.code == ValidationErrorCode::InvalidConditionFormat));
    }

    #[test]
    fn flags_duplicate_node_ids() {
        let def = graph(
            vec![node("a", NodeType::Action), node("a", NodeType::Agent)],
            vec![],
            "a",
        );
        let errors = validate_graph(&def);
        assert!(errors
            .iter()
            .any(|e| e.code == ValidationErrorCode::DuplicateNodeId));
    }
}
